### PR TITLE
Svelte Enhancements 

### PR DESCRIPTION
**Pull request by itsonlyjames**
## Todo: 
- [x] Evaluate correct leaf methods for instantiating a new sticky header: https://github.com/zhouhua/obsidian-sticky-headings/pull/25#issuecomment-2295305569. Need to better understand the Obsidian leaf, window, tab, etc. situation.
~~- [ ] Add transition to newly added/removed sticky header items.~~
 ~~- Transition in/out won't work currently as we are updating the headers on every scroll and thus rerendering the StickyHeader svelte component~~
- [x] Pre-calculate the heading height on click to scroll
  - If we have the offset of the heading, we can get the intended heading structure at that offset, and set that as the stickyHeaderHeight rather than waiting for it to update _after_ we click to scroll.
    - Initial thoughts are checking the clicked header offset, and calculating from the headings array, what elements would be shown between the highest order heading, and the closest to the clicked offset. 
  - This will also improve the `scrollBehaviour: smooth` functionality which is finicky at the moment.

### CODE REVIEW

File: src/types.d.ts [hunk 1]
```diff
@@ -24,9 +24,28 @@ export interface Heading extends HeadingCache {
...
+
   interface MarkdownSubView {
     type: 'source' | 'preview';
   }
+  interface Editor {
+    cm: {
+      scrollDOM: HTMLElement;
+      lineBlockAt: (Pos) => {};
+    };
+  }
```
Comment (zhouhua) [inline src/types.d.ts:49]: @itsonlyjames These type extensions are unnecessary. I added the obsidian-typings library, so you just need to update your dependencies.

Comment (itsonlyjames) [reply src/types.d.ts:49]: Makes sense why it was still erroring for me, let me update 👍

File: src/plugin.ts [hunk 2]
```diff
@@ -151,22 +150,44 @@ export default class StickyHeadingsPlugin extends Plugin {
...
+
+        if (heading.level < currentLevel) {
+          result.unshift(heading);
+          currentLevel = heading.level;
+        }
+
+        // Break if we've reached the highest level
+        if (currentLevel === 1) break;
+      }
+
```
Comment (zhouhua) [inline src/plugin.ts:164]: We discussed this issue in #7. The previously complex calcIndentLevels method was designed to calculate for the default mode, not just the concise mode.

Comment (itsonlyjames) [reply src/plugin.ts:164]: You're right. My initial trepidation was the seperate indent array, so let me reevaluate and combine both methods.

Comment (zhouhua) [reply src/plugin.ts:164]: We could also add an attribute to the Heading.

**Comment by zhouhua** [comment]
It seems my VSCode didn't format the code correctly, which introduced many code style issues. I'm sure this caused you a lot of trouble, and I'm very sorry about that. @itsonlyjames

File: src/plugin.ts [hunk 3]
```diff
@@ -236,6 +220,7 @@ export default class StickyHeadingsPlugin extends Plugin {
         ...heading,
         title: await parseMarkdown(heading.heading, this.app),
         offset,
+        indentLevel: heading.level - 1,
```
Comment (itsonlyjames) [inline src/plugin.ts:223]: Great thought on the property, works great 👍

**Comment by itsonlyjames** [comment]
@zhouhua we are having major conflicts between linting and formatting. I am using your created prettier.json file however this conflicts with your newly created eslint checker. Please fix as it's causing major issues going back and forth between code styles, completely redundant code commits.

**Comment by zhouhua** [comment]
I apologize for the numerous formatting issues in the code, which have caused us a lot of trouble over the past few days. I reviewed some of the problems and made several fixes. Here are the main points I want to sync with you:

1. i18n/i18n-*.ts Files: These files are auto-generated, so their formatting should be ignored. I added this setting in the Prettier and ESLint configurations.

2. ESLint and Prettier Compatibility: ESLint cannot perfectly align with Prettier, so I disabled some ESLint rules, such as semi, comma-dangle, and brace-style. If we encounter similar situations in the future (where ESLint and Prettier formatting results are inconsistent), I will continue to disable the corresponding ESLint rules. You can do the same as well.

A bit later, I will configure a Prettier pre-commit hook to prevent issues caused by related plugins not working properly in VSCode.

**Comment by itsonlyjames** [comment]
After a quick look around it seems we are both in sync, as well as Prettier and ESLint. No issues at all. 

On point number 2—I don't mind if we change the Prettier config to match ESLint, rather than disabling ESLint checks to make them compatible, it's a non-issue either way with both of them in sync. It formats correctly on save–and it lints correctly on lint now. Brilliant.

```
> obsidian-sticky-headings@1.0.6 prepare
> husky install

install command is DEPRECATED
```

Is this just a placeholder for the pre-commit hook you mentioned?

I had to run `npm i --save-dev husky` also, so I've added it to the required packages list 👍

I'm going to read through the Obsidian docs now and try to best understand the sequence of events for leaf opening, closing, and editing.

**Comment by itsonlyjames** [comment]
I've had a thought about pre-calculating the sticky heading height before scrolling to elements. See my update in the todo list comment. 

It could potentially be done here, but I'm open to ideas: https://github.com/zhouhua/obsidian-sticky-headings/pull/27/files#diff-886c646ba033ed4d87f181415d8c9d117a8a9f6e34417af98ae85c76a3d90b83R29

**Comment by itsonlyjames** [comment]
To fix: svelte files are not linting correctly and are riddled with errors

**Comment by zhouhua** [comment]
The pre-commit hook is not just a placeholder. It actually works. I followed the guide on the Prettier website to add it. However, some steps might be outdated. I'll look into it further.

Pre-calculating the sticky heading height is a very smart idea. This will make the scroll position very precise.

**Comment by zhouhua** [comment]
I implemented your idea for height prediction. There's still some room for optimization, but I feel the performance impact isn't significant since most of the additional rendering and calculation work occurs during the click event.

**Comment by itsonlyjames** [comment]
That's a shame, but I'm happy we got to test out the theory with an implementation to see if it works. 

I guess the big difference between what we're dealing with and VSCode is that we have a virtual DOM whereas VSCode doesn't. I even investigated and attempted to implement [intersection observer](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) however because of the virtual DOM this also didn't work. Need some time to think of alternate solutions.

**Comment by zhouhua** [comment]
I think there's another approach: manually setting scrollTop to simulate the animation effect. I submitted an experimental version of the code, and it seems to work quite well.

**Comment by zhouhua** [comment]
In this commit [`d250083` (#27)](https://github.com/zhouhua/obsidian-sticky-headings/pull/27/commits/d2500831efb8a6f3c08d34902d59925a2e254906), I fixed almost all the issues mentioned in this comment https://github.com/zhouhua/obsidian-sticky-headings/pull/25#issuecomment-2295305569.

**Comment by zhouhua** [comment]
@itsonlyjames, Over the past week, we've made a lot of progress. In my testing, the code on this branch is running well. I think this PR might be ready to merge into the master. What do you think?

**Comment by itsonlyjames** [comment]
I think it's super close! I'll remove the todo item for transitions as this can wait. I think we only have two things left to do, fix the issue with the leaves and file-open (I still get the wrong headers sometimes), and loading/unloading the plugin correctly so when it's updated Obsidian doesn't need to be restarted.

**Comment by zhouhua** [comment]
I might be wrong, but I think the two issues you mentioned might have already been fixed.

Regarding the 'open-file' issue, I addressed it in yesterday's commit. There were two minor problems with the previous implementation:

1. When leaf.id already existed in fileResolveMap, we skipped its subsequent processing, but the opened file might have changed.

2. When a file was already open, splitting to the right or down with the same file wouldn't trigger the 'file-open' event again. So, I replaced 'file-open' with 'active-leaf-change'.

Regarding the second issue you mentioned, my tests show everything is working fine. Whether enabling/disabling the plugin or modifying settings, there's no need to restart Obsidian.(Maybe I overlooked something?)

**Comment by itsonlyjames** [comment]
Nah I'm still getting issues with both changing file and the reloading—it's to do with the unload function; change some code and the hotreload plugin highlights this issue, it doesn't reload the component. 

If I change active file then it uses the old headings. Tests are failing for me. Doesn't work when splitting, or clicking on a new file. 

Aug-25-2024 11-16-33

**Comment by zhouhua** [comment]
I'm a bit confused. Is there a difference in the code we're using?

reload

I haven't noticed any error messages from hotreload.



And splitting the current document to the right also works normally.

**Comment by itsonlyjames** [comment]
Terribly strange.. I mean if it works fine for you then it's probably all good, I'll let you make the call on this. If you can't replicate it's probably fine

**Pull request merged**